Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -153,567 +133,757 @@ final class ConnectionPool {
-                return futureConnection
-            }
+extension Connection: Hashable {
+    static func == (lhs: Connection, rhs: Connection) -> Bool {
+        return ObjectIdentifier(lhs) == ObjectIdentifier(rhs)
+    }
+
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(ObjectIdentifier(self))
+    }
+}
+
+/// A connection provider of `HTTP/1.1` connections with a given `Key` (host, scheme, port)
+///
+/// On top of enabling connection reuse this provider it also facilitates the creation
+/// of concurrent requests as it has built-in politeness regarding the maximum number
+/// of concurrent requests to the server.
+class HTTP1ConnectionProvider: CustomStringConvertible {
+    enum Action {
+        case lease(Connection, Waiter)
+        case create(Waiter)
+        case replace(Connection, Waiter)
+        case closeProvider
+        case park(Connection)
+        case none
+        case fail(Waiter, Error)
+        case cancel(Connection, Bool)
+        indirect case closeAnd(Connection, Action)
+        indirect case parkAnd(Connection, Action)
+    }
+
+    struct ConnectionsState {
+        enum State {
+            case active
+            case closed
         }
 
-        func release(connection: Connection) {
-            self.activityPrecondition(expected: [.opened, .closing])
-            let action = self.parentPool.connectionProvidersLock.withLock {
-                self.stateLock.withLock { self.state.releaseAction(for: connection) }
+        #if DEBUG
+            struct Snapshot {
+                var state: State
+                var availableConnections: CircularBuffer<Connection>
+                var leasedConnections: Set<Connection>
+                var waiters: CircularBuffer<Waiter>
+                var openedConnectionsCount: Int
+                var pending: Int
             }
-            switch action {
-            case .succeed(let promise):
-                promise.succeed(connection)
+        #endif
 
-            case .makeConnectionAndComplete(let eventLoop, let promise):
-                self.makeConnection(on: eventLoop).cascade(to: promise)
+        let maximumConcurrentConnections: Int
+        let eventLoop: EventLoop
 
-            case .replaceConnection(let eventLoop, let promise):
-                connection.close().flatMap {
-                    self.makeConnection(on: eventLoop)
-                }.whenComplete { result in
-                    switch result {
-                    case .success(let connection):
-                        promise.succeed(connection)
-                    case .failure(let error):
-                        promise.fail(error)
-                    }
-                }
+        private var state: State = .active
 
-            case .none:
-                break
-            }
+        /// Opened connections that are available.
+        private var availableConnections: CircularBuffer<Connection> = .init(initialCapacity: 8)
```

Comment:
I think these variables should eventually move into the `State` but I'm happy if we do this in a followup.

For example if `state = .inactive`, what if `leasedConnections` isn't empty. Or `pending != 0` or something like that?

Here's an example state machine like that: https://github.com/apple/swift-nio-examples/blob/master/backpressure-file-io-channel/Sources/BackpressureChannelToFileIO/FileIOChannelWriteCoordinator.swift#L44-L81

Reply:
I though about that, but decided to leave it out for now, as this already taken way too much time, there are other things I want to improve here as well...

Reply:
@artemredkin yes, good call. We can followup